packages/notebook/src/celllist.ts

```diff
@@ -0,0 +1,552 @@
+    });
+    return this.length;
+  }
+
+  /**
+   * Insert a set of items into the cell list at the specified index.
+   *
+   * @param index - The index at which to insert the cells.
+   *
+   * @param cells - The cells to insert at the specified index.
+   *
+   * @returns The new length of the cell list.
+   *
+   * #### Complexity.
+   * Linear.
+   *
+   * #### Iterator Validity
+   * No changes.
+   *
+   * #### Notes
+   * The `index` will be clamped to the bounds of the cell list.
+   *
+   * #### Undefined Behavior.
+   * An `index` which is non-integral.
+   *
+   * #### Notes
+   * This should be considered to transfer ownership of the
+   * cells to the `CellList`. As such, `cell.dispose()` should
+   * not be called by other actors.
+   */
+  insertAll(index: number, cells: IterableOrArrayLike<ICellModel>): number {
+    let newValues = toArray(cells);
+    each(newValues, cell => {
+      // Generate a new uuid for the cell.
+      let id = utils.uuid();
+      this._cellMap.set(id, cell);
+      this._cellOrder.beginCompoundOperation();
+      this._cellOrder.insert(index++, id);
+      this._cellOrder.endCompoundOperation();
+    });
+    return this.length;
+  }
+
+  /**
+   * Remove a range of items from the cell list.
+   *
+   * @param startIndex - The start index of the range to remove (inclusive).
+   *
+   * @param endIndex - The end index of the range to remove (exclusive).
+   *
+   * @returns The new length of the cell list.
+   *
+   * #### Complexity
+   * Linear.
+   *
+   * #### Iterator Validity
+   * Iterators pointing to the first removed cell and beyond are invalid.
+   *
+   * #### Undefined Behavior
+   * A `startIndex` or `endIndex` which is non-integral.
+   */
+  removeRange(startIndex: number, endIndex: number): number {
+    this._cellOrder.removeRange(startIndex, endIndex);
+    return this.length;
+  }
+
+  /**
+   * Whether the object can redo changes.
+   */
+  get canRedo(): boolean {
+    return this._cellOrder.canRedo;
+  }
+
+  /**
+   * Whether the object can undo changes.
+   */
+  get canUndo(): boolean {
+    return this._cellOrder.canUndo;
+  }
+
+  /**
+   * Begin a compound operation.
+   *
+   * @param isUndoAble - Whether the operation is undoable.
+   *   The default is `true`.
+   */
+  beginCompoundOperation(isUndoAble?: boolean): void {
+    this._cellOrder.beginCompoundOperation(isUndoAble);
+  }
+
+  /**
+   * End a compound operation.
+   */
+  endCompoundOperation(): void {
+    this._cellOrder.endCompoundOperation();
+  }
+
+  /**
+   * Undo an operation.
+   */
+  undo(): void {
+    this._cellOrder.undo();
+  }
+
+  /**
+   * Redo an operation.
+   */
+  redo(): void {
+    this._cellOrder.redo();
+  }
+
+  /**
+   * Clear the change stack.
+   */
+  clearUndo(): void {
+    // Dispose of cells not in the current
+    // cell order.
+    for (let key of this._cellMap.keys()) {
+      if (ArrayExt.findFirstIndex(
+         toArray(this._cellOrder), id => (id)===key) === -1) {
```

Comment:
the formatting of this line is confusing, should this be `=> id === key) === -1)`?